
---
title: "Haplotype Analysis Report for *`r params$LOCUS`*"
author: Dariush Ghasemi
params:
  date: !r Sys.Date()
  LOCUS: "PDILT"
output: html_notebook
format: 
  revealjs:
    transition: slide,
    background-transition: fade
editor: visual
---

### Date: `r format(Sys.Date(), "%B %d, %Y")`

```{r, echo=FALSE}
#date: !r Sys.Date()
#Sys.Date()
```

```{r, echo=FALSE}
# date
today.date <- format(Sys.Date(), "%d-%b-%y")

regional   <- paste0("~/projects/gwas/05_regional_association/LZ_plots/09-Mar-23_", params$LOCUS, "-1.png")
density    <- paste0("output/plot_histogram/", today.date, "_plot_histo_", params$LOCUS, ".png")
annotation <- paste0("output/plot_annotations/", today.date, params$LOCUS, "_plot_annotations.png")
genes      <- paste0("output/plot_genes/", today.date, params$LOCUS, "_plot_genes.png")
haplotypes <- paste0("output/plot_haplotypes/", today.date, params$LOCUS, "_plot_haplotypes_shrinked.png")
heatmap    <- paste0("output/plot_heatmaps/", today.date, params$LOCUS, "_plot_heatmap_haplotypes_effect.png")
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
# [install the libraries before if not already installed]
# Loading libraries 
library(knitr)
library(rmarkdown)
#install.libraries("patchwork")
#install.packages("cowplot")
library(cowplot)
#library(ggplot2)
```

## Introduction
This is a report of Haplotype reconstrauction analysis for *`r params$LOCUS`* generated using RMarkdown.


## Plot Regional Association

```{r histogram, echo=FALSE, fig.cap="Density", out.width='70%'}
knitr::include_graphics(regional)
```

## Plot Variants Frequency, Annotation and Allele Frequency

```{r Variants, echo=FALSE, fig.cap="Number and annotation of exonic variants in each gene", fig.align = "default", fig.subcap=c('a','b', 'c'), out.width='70%'}
#p1 <- ggdraw() + draw_image(genes)
#p2 <- ggdraw() + draw_image(density)
#p3 <- ggdraw() + draw_image(annotation)
#plot_grid(p1, p2, p3, cnol = 2)

knitr::include_graphics(genes)
knitr::include_graphics(annotation)
knitr::include_graphics(density)
```

## Plot Haplotypes

```{r haplotypes, figures-side, echo=FALSE, fig.cap="Haplotypes", fig.show="hold", out.width='80%'}
knitr::include_graphics(haplotypes)
```

## Plot Heatmap

```{r heatmap, echo=FALSE, fig.cap="Heatmap Plot", out.width='90%'}

knitr::include_graphics(heatmap)
```
