
---
title: "Haplotype Analysis Report for *`r params$LOCUS`*"
author: Dariush Ghasemi
params:
  date: !r Sys.Date()
  LOCUS: "PDILT"
output: 
  html_notebook:
    theme: journal #united #flatly
    highlight: #espresso #tango 
    toc: TRUE
    toc_depth: 5
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      number_sections: FALSE
format: 
  revealjs:
    transition: slide,
    background-transition: fade
editor: visual
---

### Date: `r format(Sys.Date(), "%B %d, %Y")`

```{r, echo=FALSE}
#date: !r Sys.Date()
#Sys.Date()
```

```{r, echo=FALSE}
# date
today.date <- format(Sys.Date(), "%d-%b-%y")
#base.dir <- 
regional   <- paste0("~/projects/gwas/05_regional_association/LZ_plots/09-Mar-23_", params$LOCUS, "-1.png")
density    <- paste0("output/plot_histogram/old/26-Oct-23_plot_histo_", params$LOCUS, ".png")
annotation <- paste0("output/plot_annotations/old/18-Dec-23_", params$LOCUS, "_plot_annotations.png")
genes      <- paste0("output/plot_genes/old/18-Dec-23_", params$LOCUS, "_plot_genes.png")
haplotypes <- paste0("output/plot_haplotypes/", params$LOCUS, "_plot_haplotypes.png")
haplotypes.shrinked <- paste0("output/plot_haplotypes/", params$LOCUS, "_plot_haplotypes_shrinked.png")
heatmap.phen <- paste0("output/plot_heatmaps/", params$LOCUS, "_plot_heatmap_haplotypes_effect_on_phen.png")
heatmap.meta <- paste0("output/plot_heatmaps/", params$LOCUS, "_plot_heatmap_haplotypes_effect_on_meta.png")
heatmap.prot <- paste0("output/plot_heatmaps/", params$LOCUS, "_plot_heatmap_haplotypes_effect_on_prot.png")
res.sig.phen <- paste0("output/significant_result/", params$LOCUS, "_haplotypes_association_with_phen.csv")
res.sig.meta <- paste0("output/significant_result/", params$LOCUS, "_haplotypes_association_with_meta.csv")
res.sig.prot <- paste0("output/significant_result/", params$LOCUS, "_haplotypes_association_with_prot.csv")

```



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r, include=FALSE}
# [install the libraries before if not already installed]
# Loading libraries

library(knitr)
library(rmarkdown)
library(tidyverse)

#library(cowplot)
#library(ggplot2)
```




### Introduction
This is a report of Haplotype reconstrauction analysis for *`r params$LOCUS`* generated using RMarkdown.



### Plot Variants Allele Frequency

```{r Variants, echo=FALSE, fig.align = "center", out.width='70%', fig.cap="Figure 4. Allele frequency of the variants", fig.subcap=c('a','b', 'c')}

#p1 <- ggdraw() + draw_image(genes)
#p2 <- ggdraw() + draw_image(density)
#p3 <- ggdraw() + draw_image(annotation)
#plot_grid(p1, p2, p3, cnol = 2)

knitr::include_graphics(density)
```

|
|
|

### Plot Regional Association

```{r LocusZoom, echo=FALSE, fig.align="center", out.width='70%', fig.cap="Figure 1. Regional association plot of the locus"}

knitr::include_graphics(regional)
```

|
|
|

### Plot Variants Counts in each Gene

```{r genes, echo=FALSE, fig.align="center", out.width='70%', fig.cap="Figure 2. Number of the variants in each gene"}

knitr::include_graphics(genes)
```

|
|
|

### Plot Variants Annotation

```{r annotation, echo=FALSE, fig.align="center", out.width='70%', fig.cap="Figure 3. Annotation of the variants in each gene"}

knitr::include_graphics(annotation)
```

|
|
|

### Plot Haplotypes

```{r haplotypes, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 4. Haplotypes"}

knitr::include_graphics(haplotypes)
```

|
|
|

### Plot Haplotypes Narrowed

```{r haplotypes narrowed, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='70%', fig.cap="Figure 5. Haplotypes narrowed with varied variants"}

knitr::include_graphics(haplotypes.shrinked)
```

|
|
|

### Results for Clinical Traits 

#### Plot Heatmap: Haplotypes Associations with Clinical Traits

```{r heatmap traits, echo=FALSE, fig.align="center", out.width='90%', fig.cap="Figure 6. Heatmap Plot - Clinical traits"}

knitr::include_graphics(heatmap.phen)
```

|
|
|

#### Significant Associations with Clinical Traits, Metabolites and Proteins

```{r significant results traits, echo=FALSE, message=FALSE, warning=TRUE}

result_phen <- read.delim(file = res.sig.phen, header = TRUE, sep = ",", comment.char = "")

# show the results
result_phen %>% DT::datatable(caption = "Table 1. Results of haplotypes association with clinical traits")
```

|
|
|

### Results for Serum Metabolites

#### Plot Heatmap: Haplotypes Associations with Serum Metabolites

```{r heatmap metabolites, echo=FALSE, fig.align="center", out.width='120%', fig.cap="Figure 7. Heatmap Plot - Metabolites"}

knitr::include_graphics(heatmap.meta)
```

|
|
|

#### Significant Associations with Metabolites
```{r significant results metabolites, echo=FALSE, message=FALSE, warning=TRUE}

result_phen <- read.delim(file = res.sig.meta, header = TRUE, sep = ",", comment.char = "")

# show the results
result_phen %>% DT::datatable(caption = "Table 2. Results of haplotypes association with serum metabolites")
```

|
|
|

### Results for Plasma Proteins

#### Plot Heatmap: Haplotypes Associations with Proteins

```{r heatmap protein, echo=FALSE, fig.align="center", out.width='120%', fig.cap="Figure 8. Heatmap Plot - Proteins"}

knitr::include_graphics(heatmap.prot)
```



#### Significant Associations with Plasma Proteins

```{r significant results protein, echo=FALSE, message=FALSE, warning=TRUE}

result_prot <- read.delim(file = res.sig.prot, header = TRUE, sep = ",", comment.char = "")
#result_sig <- result_sig %>% arrange(p.value, haplo)

# change columns names
#names(result_sig) <- names(result_sig) %>% str_remove("X\\.+\\d\\.")

# show the results
result_prot %>% DT::datatable(caption = "Table 3. Results of haplotypes association with plasma proteins")

```



```{r res, echo=FALSE}
# Display the table

#DT::datatable(result_sig, caption = "Results Table", options = list(pageLength = 10))
#%>% knitr::kable_styling(position = "center") 
#find /scratch/mfilosi -type f -name "*.Rmd"

#knitr::kable(
#  result_sig,
  #"simple",
  #"latex",
  #digits = 5,
  #format.args = list(scientific = TRUE), 
#  full_width = FALSE,
#  html_font = "Cambria", 
#  caption = "Results Table of Association with Clinical Traits"
#)

```