
---
title: "Haplotype Analysis Report for *`r params$LOCUS`*"
author: Dariush Ghasemi
params:
  date: !r Sys.Date()
  LOCUS: "o"
  ASSAY: "o"
  zoom: "o"
  hist: "o"
  anot: "o"
  gene: "o"
  hap1: "o"
  hap2: "o"
  heat: "o"
  res:  "o"
  summ: "o"
  
output: 
  html_notebook:
    theme: journal #united #flatly
    #highlight: #espresso #tango 
    toc: TRUE
    toc_depth: 5
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      number_sections: FALSE
format: 
  revealjs:
    transition: slide,
    background-transition: fade
editor: visual
geometry: margin=0.95in
---

### Date: `r format(Sys.Date(), "%B %d, %Y")`

```{r, echo=FALSE}
#date: !r Sys.Date()
#Sys.Date()
```

Workig directory is: `r getwd()`

```{r, echo=FALSE}
# date
today.date <- format(Sys.Date(), "%d-%b-%y")

```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, include=FALSE}
# [install the libraries before if not already installed]
# Loading libraries

library(knitr)
library(rmarkdown)
library(tidyverse)

#library(cowplot)
#library(ggplot2)
```


### Introduction
This is a report of haplotype reconstruction analysis for *`r params$LOCUS`* generated using RMarkdown.


|
|
|

### Plot Regional Association

```{r LocusZoom, echo=FALSE, fig.align="center", out.width='100%', fig.cap="Figure 1. Regional association plot of the locus"}

knitr::include_graphics(params$zoom)
```

|
|
|

### Plot Variants Allele Frequency

```{r Variants, echo=FALSE, fig.align = "center", out.width='70%', fig.cap="Figure 2. Allele frequency of the variants", fig.subcap=c('a','b', 'c')}

#p1 <- ggdraw() + draw_image(genes)
#p2 <- ggdraw() + draw_image(density)
#p3 <- ggdraw() + draw_image(annotation)
#plot_grid(p1, p2, p3, cnol = 2)

knitr::include_graphics(params$hist)
```

|
|
|

### Plot Variants Counts in each Gene

```{r genes, echo=FALSE, fig.align="center", out.width='80%', fig.cap="Figure 3. Number of the variants in each gene"}

knitr::include_graphics(params$gene)
```

|
|
|

### Plot Variants Annotation

```{r annotation, echo=FALSE, fig.align="center", out.width='80%', fig.cap="Figure 4. Annotation of the variants in each gene"}

knitr::include_graphics(params$anot)
```

|
|
|

#### Discriptive statistics 

```{r summary of the datasets, echo=FALSE, message=FALSE, warning=TRUE}

data_summ <- read.delim(file = params$summ, header = TRUE, sep = "\t", comment.char = "")

data_summ %>% knitr::kable("pipe", caption = "Table 1. Discriptive statistics of merged datasets")
```

|
|
|

### Results for Clinical Traits 

#### Plot Haplotypes

```{r haplotypes, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 5. Alleles distribution across haplotypes"}

knitr::include_graphics(str_split(params$hap1, " ", simplify = TRUE)[,1])
```

|
|
|

#### Plot Haplotypes Narrowed

```{r haplotypes narrowed, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 6. Haplotypes narrowed by varied variants"}

knitr::include_graphics(str_split(params$hap2, " ", simplify = TRUE)[,1])
```

|
|
|

#### Plot Heatmap: Haplotypes Associations with Clinical Traits

```{r heatmap traits, echo=FALSE, fig.align="center", out.width='100%', fig.cap="Figure 7. Heatmap plot of haplotypes association with clinical traits"}

knitr::include_graphics(str_split(params$heat, " ", simplify = TRUE)[,1])
```

|
|
|

#### Significant Associations with Clinical Traits, Metabolites and Proteins

```{r significant results traits, echo=FALSE, message=FALSE, warning=TRUE}

res_phen    <- str_split(params$res, " ", simplify = TRUE)[,1]
#result_phen <- read.delim(file = res_phen, header = TRUE, sep = ",", comment.char = "")
result_phen <- readRDS(res_phen)

# show the results
result_phen %>% DT::datatable(caption = "Table 2. Results of haplotypes association with clinical traits")
```

|
|
|

### Results for Serum Metabolites

#### Plot Haplotypes

```{r haplotypes metabol, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 8. Haplotypes in the subset with metabolites"}

#knitr::include_graphics(str_split(params$hap1, " ", simplify = TRUE)[,2])
```

|
|
|

#### Plot Haplotypes Narrowed

```{r haplotypes narrowed metabol, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='70%', fig.cap="Figure 9. Haplotypes narrowed by varied variants in the subset with metabolites"}

knitr::include_graphics(str_split(params$hap2, " ", simplify = TRUE)[,2])
```

|
|
|

#### Plot Heatmap: Haplotypes Associations with Serum Metabolites

```{r heatmap metabolites, echo=FALSE, fig.align="center", out.width='120%', fig.cap="Figure 10. Heatmap plot of haplotypes association with Metabolites"}

knitr::include_graphics(str_split(params$heat, " ", simplify = TRUE)[,2])
```

|
|
|

#### Significant Associations with Metabolites
```{r significant results metabolites, echo=FALSE, message=FALSE, warning=TRUE}

res_meta    <- str_split(params$res, " ", simplify = TRUE)[,2]
#result_meta <- read.delim(file = res_meta, header = TRUE, sep = ",", comment.char = "")
result_meta <- readRDS(res_meta)

# show the results
result_meta %>% DT::datatable(caption = "Table 3. Results of haplotypes association with serum metabolites")
```

|
|
|

### Results for Plasma Proteins

#### Plot Haplotypes

```{r haplotypes prot, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 11. Haplotypes in the subset with proteins"}

#knitr::include_graphics(str_split(params$hap1, " ", simplify = TRUE)[,3])
```

|
|
|

#### Plot Haplotypes Narrowed

```{r haplotypes narrowed prot, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='70%', fig.cap="Figure 12. Haplotypes narrowed by varied variants in the subset with proteins"}

knitr::include_graphics(str_split(params$hap2, " ", simplify = TRUE)[,3])
```

|
|
|

#### Plot Heatmap: Haplotypes Associations with Proteins

```{r heatmap protein, echo=FALSE, fig.align="center", out.width='120%', fig.cap="Figure 13. Heatmap plot of haplotypes association with proteins"}

knitr::include_graphics(str_split(params$heat, " ", simplify = TRUE)[,3])
```

|
|
|

#### Significant Associations with Plasma Proteins

```{r significant results protein, echo=FALSE, message=FALSE, warning=TRUE}

res_prot    <- str_split(params$res, " ", simplify = TRUE)[,3]
#result_prot <- read.delim(file = res_prot, header = TRUE, sep = ",", comment.char = "")
result_prot <- readRDS(res_prot)

#result_sig <- result_sig %>% arrange(p.value, haplo)

# change columns names
##names(result_sig) <- names(result_sig) %>% str_remove("X\\.+\\d\\.")

# show the results
result_prot %>% DT::datatable(caption = "Table 4. Results of haplotypes association with plasma proteins")

```



```{r res, echo=FALSE}
# Display the table

#DT::datatable(result_sig, caption = "Results Table", options = list(pageLength = 10))
#%>% knitr::kable_styling(position = "center") 
#find /scratch/mfilosi -type f -name "*.Rmd"

#knitr::kable(
#  result_sig,
  #"simple",
  #"latex",
  #digits = 5,
  #format.args = list(scientific = TRUE), 
#  full_width = FALSE,
#  html_font = "Cambria", 
#  caption = "Results Table of Association with Clinical Traits"
#)

```